runtime/compiler/control/HookedByTheJit.cpp

```diff
@@ -4429,6 +4429,163 @@ void JitShutdown(J9JITConfig * jitConfig)
    TRC_JIT_ShutDownEnd(vmThread, "end of JitShutdown function");
    }
 
+#if defined (LINUX)
+//   Since Linux 4.0 only users with the CAP_SYS_ADMIN capability can get PFNs.
+//   In 4.0 and 4.1 opens by unprivileged fail with -EPERM.  Starting from
+//   4.2 the PFN field is zeroed if the user does not have CAP_SYS_ADMIN.
+//   Reason: information about PFNs helps in exploiting Rowhammer vulnerability.
+typedef struct __attribute__ ((__packed__))
+   {
+   union
+      {
+      uint64_t _pmd; // page memory descriptor
+      uint64_t _pageFrameNumber : 55; // bits 0..54 represent the page frame number (PFN) if present
+      struct
+         {
+         uint64_t _swapType: 5;    // swap type if swapped
+         uint64_t _swapOffset: 50; // swap offset if swapped
+         uint64_t _softDirty: 1;   // bit 55: pte is soft dirty
+         uint64_t _exclusive: 1;   // bit 56: page exclussively mapped
+         uint64_t _zero: 4;        // bits 57..60 are zero
+         uint64_t _filePage: 1;    // bit 61: page is file-page or shared-anon
+         uint64_t _swapped: 1;     // bit 62: page is swapped
+         uint64_t _present: 1;     // bit 63: page is present
+         };
+       };
+   } pmd_t;
+
+// The following function is used to count the number of pages from a given
+// memory range that are present in physical memory, swapped or file-paged
+// It is used for diagnostics only since it has a large runtime overhead
+#include <fcntl.h> // open, O_RDONLY
+int countPresentPages(uintptr_t startAddr, uintptr_t endAddr, int &swappedCount, int &filePageCount)
+   {
+   int presentCount = 0;
+   size_t pageSize = sysconf(_SC_PAGESIZE);
+   if (startAddr >= endAddr)
+      return -1;
+
+   static const char *pagemapPath = "/proc/self/pagemap";
+
+   int pagemapfd = open(pagemapPath, O_RDONLY);
+   if (pagemapfd < 0)
+      {
+      perror("cannot open pagemap file");
+      return -1;
+      }
+   for (uintptr_t i = startAddr; i < endAddr; i += pageSize)
+      {
+      pmd_t pmd;
+      // read at given offset
+      if (pread(pagemapfd, &pmd._pmd, sizeof(pmd._pmd), (i / pageSize) * sizeof(pmd)) != sizeof(pmd))
+         {
+         perror("cannot read from pagemap file");
+         close(pagemapfd);
+         return -1;
+         }
+      if (pmd._pmd)
+         {
+         if (pmd._present)
+            {
+            presentCount++;
+            }
+         else
+            {
+            if (pmd._swapped)
+               swappedCount++;
+            if (pmd._filePage)
+               filePageCount++;
+            }
+         }
+      }
+    close(pagemapfd);
+    return presentCount;
+   }
+#endif // LINUX
+
+int countPresentPagesInSegment(const J9MemorySegment *segment, int &swappedCount, int &filePageCount)
+   {
+   return countPresentPages((uintptr_t)segment->heapBase, (uintptr_t)segment->heapTop, swappedCount, filePageCount);
```

Comment:
This should be guarded with an ifdef since `countPresentPages` is only defined under the ifdef above.

Reply:
Done